Sixth review

Cosmin-Crisan opened the pull request:
Added a new interface called `DataManager `and moved all the relevant methods to `manageData`

Cosmin-Crisan committed:
Added new Interface called DataManager

Cosmin-Crisan committed:
Moved managePracticeData to manageData

Cosmin-Crisan committed:
Update ManageArray.java

Added some TODO's

danalexandrunicoara commented on src/bussinessLogic/ManageArray.java:
```diff
@@ -42,35 +39,12 @@ public ManageArray(MapManager manageMap) {
-    private void setCharArray(int alphabetMultiplier) {
-        this.alphabetMultiplier = alphabetMultiplier;
+    public void setEvaluationArray(int alphabetMultiplier) {
         evaluationArray = new char[numberOfLetters * alphabetMultiplier];
```
This initialization can be done when declaring this property at class level.

Cosmin-Crisan replied:
👍

danalexandrunicoara commented on src/bussinessLogic/ManageArray.java:
```diff
@@ -137,6 +112,7 @@ private void setPracticeArray() {
+        //TODO - add comment here
         for (int i = 0; i < charArrayFromSortedMap.length; i++) {
             for (int j = 0; j <= i; j++) {
                 practiceArray[index] = charArrayFromSortedMap[i];
```
_index_ has the same value as _i_, from the first _for_ look. It means it can be easily replaced with it.

Cosmin-Crisan replied:
👍

Cosmin-Crisan replied:
Wrong. _Index_ does not change at the same pace with _i_. _Index_ has to cover the length of `practiceArray` which is different from `charArrayFromSortedMap` (which is covered by _i_)

danalexandrunicoara commented on src/bussinessLogic/ManageArray.java:
```diff
@@ -145,6 +121,14 @@ private void setPracticeArray() {
+     * return the evaluation array
+     */
+    @Override
+    public char[] getEvaluationArray() {
```
👍

danalexandrunicoara commented on src/bussinessLogic/ManageArray.java:
```diff
@@ -153,5 +137,11 @@ private void setPracticeArray() {
+     * return the practice array
+     */
+    @Override
+    public char[] getPracticeArray(){return practiceArray;}
```
👍

danalexandrunicoara commented on src/bussinessLogic/ManageData.java:
```diff
@@ -0,0 +1,40 @@
+     * prepare all the data necessary for the evaluation phase
+     */
+    public void manageEvaluationData() {
+        manageArray.setEvaluationArray(alphabetMultiplier);
```
use _this_ keyword to know that this is a reference to a private property.

Cosmin-Crisan replied:
👍 
This is also valid for `managePracticeData()` method, am I right?

danalexandrunicoara replied:
Yes.

danalexandrunicoara commented on src/bussinessLogic/ManageData.java:
```diff
@@ -0,0 +1,40 @@
+    private final ArrayManager manageArray;
+    // int for calculating how many times a letter is displayed in the evaluation
+    // sequence
+    private int alphabetMultiplier = 1;
```
I have seen that there are multiple constants in several classes inside this project.
An idea would be to create a new class inside which these constants could be moved and then, referenced throughout the project.
This way there will be _one single point of truth_ regarding those.

At the moment one of the constants is duplicated. See **ManageMap** class and **numberOfLetters** property together with **ManageArray** class and **numberOfLetters** property.

Cosmin-Crisan replied:
> An idea would be to create a new class inside which these constants could be moved and then, referenced throughout the project.
> This way there will be one single point of truth regarding those.

That's exactly my thought also. This is the way I was planning to solve the initialization problem on the CBM program :grinning:

danalexandrunicoara commented on src/bussinessLogic/ManageData.java:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * prepare all the data necessary for the practice phase
+     */
+    public void managePracticeData() {
```
Inside this method, there are several operations that apply for the character list. An idea here would be to use _Fluent Interface design pattern_.

Cosmin-Crisan replied:
Thanks! I'll apply this advice in a further pull request.

danalexandrunicoara commented on src/interfaces/DataManager.java:
```diff
@@ -0,0 +1,20 @@
+package interfaces;
```
👍

danalexandrunicoara commented on src/main/StartUp.java:
```diff
@@ -1,7 +1,9 @@
 package main;
```
Inside the method _runProgram_, the first line of code is: **headerLabel.setText(String.valueOf(array[charArrayIndex]));**. This should be moved inside _prepareGUI_.

Cosmin-Crisan replied:
This line of code sets and displays the character that has to be typed. If  I move it to _prepareGUI_ I'll have to implement a method that is called every time a new character is displayed. Or do you have other suggestions?

danalexandrunicoara replied:
Yes, that would be a way.

I see now that _headerLabel_ property is called several times inside _runProgram_ method. You could leave the code as is here.

Cosmin-Crisan replied:
👍

danalexandrunicoara commented on src/main/StartUp.java:
```diff
@@ -12,8 +14,10 @@
 import java.time.Instant;
 
 public class StartUp {
```
I don't think there is a need for a constructor in this class. The program can start from **main(String[] args)** method.

Cosmin-Crisan replied:
You're saying to call the `prepareGUI() `and `runEvaluation()` methods directly from `main`?

danalexandrunicoara replied:
Yes.

Cosmin-Crisan replied:
👍

Cosmin-Crisan committed:
Added the changes from the resolved conversations

The changes from the other comments will be made in another pull request for added clarity
